src/runner.ts

```diff
@@ -0,0 +1,578 @@
+    this._tests = options.tests || [];
+    this._verbose = !!options.verbose;
+    this._wctFlags = options.wctFlags ? options.wctFlags.join(' ') : '';
+    this._workspace = {
+      dir: (options.workspaceDir || './tattoo_workspace'),
+      repos: new Map()
+    };
+
+    // TODO(usergenic): Rate limit should be an option.
+    this._testRateLimiter = new Bottleneck(1, 100);
+
+    if (this._verbose) {
+      console.log('Tattoo Runner configuration:');
+      console.log({
+        excludeRepos: this._excludeRepos,
+        fresh: this._fresh,
+        repos: this._repos,
+        skipTests: this._skipTests,
+        tests: this._tests,
+        wctFlags: this._wctFlags,
+        workspaceDir: this._workspace.dir
+      });
+    }
+  }
+
+  /**
+ * Analyzes all of the HTML in 'repos/*' with hydrolysis.
+ *
+ * @returns a promise of the hydrolysis.Analyzer with all of the info loaded.
+ */
+  async _analyzeRepos() {
+    const dirs = fs.readdirSync(this._workspace.dir);
+    const htmlFiles: string[] = [];
+
+    for (const dir of dirs) {
+      for (const fn of fs.readdirSync(path.join('repos', dir))) {
+        if (/index\.html|dependencies\.html/.test(fn) ||
+            !fn.endsWith('.html')) {
+          continue;
+        }
+        // We want to ignore files with 'demo' in them, unless the element's
+        // directory has the word 'demo' in it, in which case that's
+        // the whole point of the element.
+        if (!/\bdemo\b/.test(dir) && /demo/.test(fn)) {
+          continue;
+        }
+        htmlFiles.push(path.join('repos', dir, fn));
+      }
+    }
+
+    function filter(repo: string) {
+      return !util.existsSync(repo);
+    }
+
+    // This code is conceptually simple, it's only complex due to ordering
+    // and the progress bar. Basically we call analyzer.metadataTree on each
+    // html file in sequence, then finally call analyzer.annotate() and return.
+    const analyzer = await hydrolysis.Analyzer.analyze(
+        path.join(this._workspace.dir, 'polymer', 'polymer.html'), {filter});
+
+    const progressBar =
+        util.standardProgressBar('Analyzing...', htmlFiles.length + 1);
+
+    for (const htmlFile of htmlFiles) {
+      await analyzer.metadataTree(htmlFile);
+      progressBar.tick(
+          {msg: util.progressMessage(`Analyzing ${htmlFile.slice(6)}`)});
+    }
+
+    progressBar.tick(
+        {msg: util.progressMessage('Analyzing with hydrolysis...')});
+    analyzer.annotate();
+    return analyzer;
+  }
+
+  /**
+   * Given all the repos defined in the workspace, lets iterate through them
+   * and either clone them or update their clones and set them to the specific
+   * refs.
+   */
+  async _cloneOrUpdateWorkspaceRepos() {
+    if (this._verbose) {
+      console.log('Cloning/updating workspace repos...');
+    }
+
+    const promises: Promise<nodegit.Repository>[] = [];
+
+    // Clone git repos.
+    for (const name of this._workspace.repos.keys()) {
```

Comment:
I've been trying to be supportive of node 4 so I painfully avoid array and object destructuring.